ethcontract-common/src/artifact/hardhat.rs

```diff
@@ -0,0 +1,290 @@
+use crate::contract::Network;
+use crate::errors::ArtifactError;
+use crate::{Address, Contract};
+use serde::Deserialize;
+use serde_json::Value;
+use std::collections::HashMap;
+use std::fs::File;
+use std::path::Path;
+
+/// Loads hardhat artifacts generated via `--export` and `--export-all`.
+///
+/// # Limitations
+///
+/// In hardhat, a contract could have different ABIs on different networks.
+/// This could happen when deploying test versions of contracts.
+/// Ethcontract does not support this. Parsing such artifact will result
+/// in an error. You'll have to rename contracts, or filter out networks
+/// with [`networks_allow_list`].
+///
+/// Another limitation is that hardhat allows having multiple networks
+/// with the same chain ID. For example, you can have `rinkeby`
+/// and `rinkeby-testing`. Both have chain ID of `4`, but contract addresses
+/// and ABIs can be different. Ethcontract does not support this, so you'll
+/// have to filter such networks. See [#545] for more info.
+///
+/// [#545]: https://github.com/gnosis/ethcontract-rs/issues/545.
+pub struct HardHatLoader {
+    /// Override for artifact's origin. If `None`, origin
+    /// will be derived automatically.
+    pub origin: Option<String>,
+
+    /// List of allowed network names and chain IDs.
+    ///
+    /// When loading a contract, networks with names that aren't found
+    /// in this list will be completely ignored. Contracts from these networks
+    /// will not be loaded. You can use this mechanism to bypass
+    /// the requirement that a contract must have the same ABI on all networks.
+    ///
+    /// Empty list means that all networks are allowed.
+    pub networks_allow_list: Vec<String>,
+
+    /// List of denied network names and chain IDs.
+    ///
+    /// When loading a contract, networks with names that are found
+    /// in this list will be completely ignored.
+    ///
+    /// Empty list means that no networks are denied.
+    pub networks_deny_list: Vec<String>,
+}
+
+/// Artifact format.
+#[derive(Copy, Clone, Debug)]
+pub enum Format {
+    /// Contracts for a single network. Generated with `hardhat export`.
+    SingleExport,
+
+    /// Contracts for all networks. Generated with `hardhat export --export-all`.
+    MultiExport,
+}
+
+impl HardHatLoader {
+    /// Create a new hardhat loader.
+    pub fn new() -> Self {
+        HardHatLoader {
+            origin: None,
+            networks_deny_list: Vec::new(),
+            networks_allow_list: Vec::new(),
+        }
+    }
+
+    /// Create a new hardhat loader and set an override for artifact's origins.
+    pub fn with_origin(origin: impl Into<String>) -> Self {
+        HardHatLoader {
+            origin: Some(origin.into()),
+            networks_deny_list: Vec::new(),
+            networks_allow_list: Vec::new(),
+        }
+    }
+
+    /// Set new override for artifact's origin. See [`origin`] for more info.
+    ///
+    /// [`origin`]: #structfield.origin
+    #[inline]
```

Comment:
I don't strongly mind this but I feel that specifying inline is unnecessary. I feel it is better to leave it up to the compiler unless it is performance critical which I don't imagine for this struct.

Reply:
Rust still doesn't support inlining non-`#[inline]` functions across crates. That's fine if users of our library enable lto, but we can't guarantee this.

Reply:
This certainly needs a discussion, but I'm ok with removing inline hints here.

Reply:
What I mean is that I doubt inlining this matters compared to doing the much more expensive task of deserializing json afterwards. From my side no need to remove but wanted to mention it.

Reply:
Personally I'm OK either way. I think inlining makes perfect sense for these functions (and would be surprised if the LTO optimizer didn't do it even without the attribute).

That being said I also see @vkgnosis's point that it isn't strictly necessary as artifact parsing is probably not going to be in a performance critical section of code.

My personal opinion is that since its there and I believe its correct, we should leave it. However, I don't think in general we need to be strict with our use of `#[inline]` unless we are doing some performance critical code where inlining some function that is used in a tight loop makes a significant impact (and I would hope we would have benchmakrs in those cases to prove its usefulness).

Reply:
Since there are no other inline hints in the project, I'll remove them just to keep code style uniform. I agree that they don't do much difference, and it's compiler task to optimise these things. I agree that we shouldn't be strict about them.